### PR TITLE
Propagate treq request ID as a header

### PR DESCRIPTION
**Pull request by cyli**
This addresses part of #843 and gives us a treq transaction ID to trace additional log messages from `cloud_client` logs, which will be done as part of #1635.

### CODE REVIEW

**Comment by manishtomar** [comment]
Can we have a test that calls treq with headers and ensures that existing header is not changed before sending?

**Comment by cyli** [comment]
@manishtomar Yes, good idea, thanks!

File: otter/util/logging_treq.py [hunk 1]
```diff
@@ -24,16 +24,23 @@ def _log_request(treq_call, url, **kwargs):
...
     log = kwargs.pop('log', None)
+
     if not log:
         log = default_log
     method = kwargs.get('method', treq_call.__name__)
 
-    treq_transaction = str(uuid4())
+    kwargs.setdefault('headers', {})
+    treq_transaction = kwargs['headers'].setdefault('x-otter-request-id',
+                                                    [str(uuid4())])[0]
```
Comment (manishtomar) [inline otter/util/logging_treq.py]: Non-blocking nitpick: This is a little confusing. How about
```python
kwargs.setdefault('headers', {})
treq_transaction = str(uuid.uuid4())
kwargs['headers'].setdefault('x-otter-request-id', [treq_transaction])
```

File: otter/util/logging_treq.py [hunk 2]
```diff
@@ -24,16 +24,23 @@ def _log_request(treq_call, url, **kwargs):
...
+
     log = log.bind(system='treq.request', url=url, method=method,
                    url_params=kwargs.get('params'),
                    treq_request_id=treq_transaction)
     start_time = clock.seconds()
 
+    kwargs.setdefault('headers', {})
+    kwargs['headers'].setdefault('x-otter-request-id', [treq_transaction])
```
Comment (manishtomar) [inline otter/util/logging_treq.py]: This is already set above right? Why is this set again? Sorry, I might be missing something here.

File: otter/util/logging_treq.py (hunk 1 above)
Comment (cyli) [inline otter/util/logging_treq.py]: The intention was to actually uses the request ID as the treq transaction ID, if it is provided (if we wanted to overload it.

So possibly:

```python
kwargs.setdefault('headers', {})
treq_transaction = str(uuid.uuid4())
treq_transaction = kwargs['headers'].setdefault('x-otter-request-id', [treq_transaction])[0]
```

Alternately, we can just always overwrite this header, no matter what the user provides (I can't think of a case where we might want to respect this ID after all).

File: otter/util/logging_treq.py (hunk 2 above)
Comment (cyli) [inline otter/util/logging_treq.py]: Oh that's weird - this should have been deleted.  I must have messed up my conflict resolution from stashing, thanks!

File: otter/util/logging_treq.py (hunk 1 above)
Comment (manishtomar) [inline otter/util/logging_treq.py]: Oh. Didn't realize about the update. I am fine either way. The above code that honors user given id or ignoring it. Any one is fine as long as it is clear :) I am leaning towards ignoring user provided for now.

Comment (cyli) [inline otter/util/logging_treq.py]: Sounds good - if we actually come up with a use case we can try to preserve it.  Will remove this and add in a test :D

**Comment by manishtomar** [comment]
LGTM! Merging :)

**Pull request merged**